examples/DataApi-Voting-poll/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -0,0 +1,360 @@
+    val validateAddress = proofs
+      .map(_.id)
+      .toList
+      .traverse(_.toAddress[IO])
+      .map(validateProvidedAddress(_, update.owner))
+
+    val validations = createPollValidations(update, state, None)
+    for {
+      validatedAddress <- validateAddress
+      validatedPoll <- validations
+    } yield validatedAddress.productR(validatedPoll)
+  }
+
+  def voteInPollValidationsWithSignature(update: VoteInPollUpdate, proofs: NonEmptySet[SignatureProof], state: State)(implicit sp: SecurityProvider[IO]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateAddress = proofs
+      .map(_.id)
+      .toList
+      .traverse(_.toAddress[IO])
+      .map(validateProvidedAddress(_, update.address))
+
+    val validations = voteInPollValidations(update, state, None)
+
+    for {
+      validatedAddress <- validateAddress
+      validatedPoll <- validations
+    } yield validatedAddress.productR(validatedPoll)
+  }
+}
+
+object Combiners {
+  def combineCreatePoll(poll: CreatePollUpdate, acc: State): State = {
+    val pollId = Hash.fromBytes(customUpdateSerialization(poll)).toString
+    val pollOptions = poll.pollOptions.flatMap(option => Map(option -> 0L)).toMap
+    val newState = Poll(pollId, poll.name, poll.owner, pollOptions, Map.empty, poll.startSnapshotOrdinal, poll.endSnapshotOrdinal)
+
+    acc.focus(_.polls).modify(_.updated(pollId, newState))
+  }
+
+  def combineVoteInPoll(voteInPoll: VoteInPollUpdate, acc: State): State = {
+    val currentState = acc.polls(voteInPoll.pollId)
+    val currentOptionNumber = currentState.pollOptions(voteInPoll.option)
+
+    val newState = currentState
+      .focus(_.pollOptions)
+      .modify(_.updated(voteInPoll.option, currentOptionNumber + 1))
+      .focus(_.votes)
+      .modify(_.updated(voteInPoll.address, Map(voteInPoll.option -> 1)))
+
+    acc.focus(_.polls).modify(_.updated(voteInPoll.pollId, newState))
+  }
+}
+
+object Utils {
+  def customUpdateSerialization(update: PollUpdate): Array[Byte] = {
+    println("Serialize UPDATE event received")
+    println(update.asJson.deepDropNullValues.noSpaces)
+    update.asJson.deepDropNullValues.noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  def customStateSerialization(state: State): Array[Byte] = {
+    println("Serialize STATE event received")
+    println(state.asJson.deepDropNullValues.noSpaces)
+    state.asJson.deepDropNullValues.noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  def customStateDeserialization(bytes: Array[Byte]): Either[Throwable, State] = {
+    parser.parse(new String(bytes, StandardCharsets.UTF_8)).flatMap { json =>
+      json.as[State]
+    }
+  }
+
+  def customUpdateDeserialization(bytes: Array[Byte]): Either[Throwable, PollUpdate] = {
+    parser.parse(new String(bytes, StandardCharsets.UTF_8)).flatMap { json =>
+      json.as[PollUpdate]
+    }
+  }
+}
+
+object Data {
```

Comment:
I would preserve this portion as the main object in this Main method here, since it most closely encapsulates the logic repeated elsewhere, and move everything else to different utility / error / validation files.